tests/net/ieee802154/l2/src/ieee802154_test.c

```diff
@@ -444,36 +444,89 @@ static bool test_raw_packet_sending(void)
-		0x0d, 0xb8, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff,
-		0x16, 0xf0, 0x02, 0xff, 0x16, 0xf0, 0x12, 0xff, 0x16, 0xf0, 0x32,
-		0xff, 0x16, 0xf0, 0x00, 0xff, 0x16, 0xf0, 0x00, 0xff, 0x16
+		0x61, 0xd8, 0x16, 0xcd, 0xab, 0xff, 0xff, 0x00, 0x12, 0x4b, 0x00, 0x1c, 0xa1, 0xac,
+		0xe5, 0x7b, 0x39, 0x3a, 0x02, 0x01, 0xff, 0x4b, 0x12, 0x00, 0x87, 0x00, 0x0d, 0xa4,
+		0x00, 0x00, 0x00, 0x00, 0xfe, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x70, 0x14,
+		0xa6, 0x1c, 0x00, 0x4b, 0x12, 0x00, 0x01, 0x02, 0xe5, 0xac, 0xa1, 0x1c, 0x00, 0x4b,
+		0x12, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	};
+	/* Expected uncompressed IPv6 payload. */
+	static uint8_t expected_rx_pkt[] = {
+		0x60, 0x00, 0x00, 0x00, 0x00, 0x28, 0x3a, 0xff, 0xfe, 0x80, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0xe7, 0xac, 0xa1, 0x1c, 0x00, 0x4b, 0x12, 0x00, 0xff, 0x02, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0xff, 0x4b, 0x12, 0x00, 0x87, 0x00,
+		0x0d, 0xa4, 0x00, 0x00, 0x00, 0x00, 0xfe, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x70, 0x14, 0xa6, 0x1c, 0x00, 0x4b, 0x12, 0x00, 0x01, 0x02, 0xe5, 0xac, 0xa1, 0x1c,
+		0x00, 0x4b, 0x12, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	};
+	struct sockaddr_ll recv_src_sll = {0};
+	struct sockaddr_ll socket_sll = {
+		.sll_ifindex = net_if_get_by_iface(iface),
+		.sll_family = AF_PACKET,
+		.sll_protocol = ETH_P_IEEE802154,
+	};
+	uint8_t received_payload[80] = {0};
+	struct timeval timeo_optval = {
+		.tv_sec = 1,
+		.tv_usec = 0,
 	};
 	struct ieee802154_mpdu mpdu;
-	struct net_pkt *pkt;
+	socklen_t recv_src_sll_len;
 	struct net_buf *frag;
+	struct net_pkt *rx_pkt;
+	bool result = false;
+	int received_len;
+	int fd;
 
 	NET_INFO("- Sending ACK reply to a data packet\n");
 
-	pkt = net_pkt_rx_alloc(K_FOREVER);
-	frag = net_pkt_get_frag(pkt, sizeof(data_pkt), K_FOREVER);
+	fd = socket(AF_PACKET, SOCK_DGRAM, ETH_P_IEEE802154);
+	if (fd < 0) {
+		NET_ERR("*** Failed to create DGRAM socket : %d\n", errno);
+		goto out;
+	}
+
+	if (bind(fd, (const struct sockaddr *)&socket_sll, sizeof(struct sockaddr_ll))) {
+		NET_ERR("*** Failed to bind packet socket : %d\n", errno);
+		goto release_fd;
+	}
+
+	if (setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeo_optval, sizeof(timeo_optval))) {
+		NET_ERR("*** Failed to set reception timeout on packet socket : %d\n", errno);
+		goto release_fd;
+	}
+
+	rx_pkt = net_pkt_rx_alloc(K_FOREVER);
+	frag = net_pkt_get_frag(rx_pkt, sizeof(data_pkt), K_FOREVER);
 
 	memcpy(frag->data, data_pkt, sizeof(data_pkt));
 	frag->len = sizeof(data_pkt);
 
-	net_pkt_frag_add(pkt, frag);
+	net_pkt_frag_add(rx_pkt, frag);
 
-	if (net_recv_data(iface, pkt) < 0) {
+	if (net_recv_data(iface, rx_pkt) < 0) {
 		NET_ERR("Recv data failed");
-		return false;
+		goto release_rx_pkt;
+	}
+
+	recv_src_sll_len = sizeof(recv_src_sll);
+	received_len = recvfrom(fd, received_payload, sizeof(received_payload), 0,
+				(struct sockaddr *)&recv_src_sll, &recv_src_sll_len);
```

Comment:
Could we also check the various values in recv_src_sll? This is kind of socket test but it does not harm to do it here too, especially as the ll structs for 802.15.4 are probably not tested anywhere else.

Reply:
Yep, good idea. Fixed.